### PR TITLE
Use ZipDeploy API when deploying functions apps 

### PR DESCRIPTION
**Pull request by dannysongg**
## Description

Adding logic to use ZipDeploy instead of OneDeploy when deploying functions apps. This change also adds a switch to deploy using ZipDeploy.

## Mandatory Checklist

- Please choose the target release of Azure PowerShell
  - [X] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] N/A

- [X] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs [hunk 1]
```diff
@@ -86,74 +89,93 @@ public override void ExecuteCmdlet()
...
-                throw new Exception("Unknown artifact type.");
-            }
+                paramValues.Add("path", TargetPath);
 
-            paramValues.Add("path", TargetPath);
+                // default async to true if not provided to match old behavior
+                if (Async.IsPresent)
+                {
+                    paramValues.Add("async", Async.ToString());
+                }
```
Comment (patelchandni) [inline src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:149]: NIT: remove empty line between if else

File: src/Websites/Websites/help/Publish-AzWebApp.md [hunk 2]
```diff
@@ -15,15 +15,17 @@ Deploys an Azure Web App from a ZIP, JAR, or WAR file using zipdeploy.
 ### FromWebApp (Default)
 ```
 Publish-AzWebApp -ArchivePath <String> [-Type <String>] [-Clean] [-Async] [-Restart] [-TargetPath <String>]
- [-IgnoreStack] [-Reset] [-Force] [-AsJob] [-Timeout <Double>] [-WebApp] <PSSite>
- [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [<CommonParameters>]
+ [-IgnoreStack] [-Reset] [-UseZipDeploy] [-Force] [-AsJob] [-Timeout <Double>] [-WebApp] <PSSite>
+ [-DefaultProfile <IAzureContextContainer>] [-ProgressAction <ActionPreference>] [-WhatIf] [-Confirm]
```
Comment (patelchandni) [inline src/Websites/Websites/help/Publish-AzWebApp.md:19]: I think ProgressAction is removed on "main" branch. Was this added back intentionally or just need to rebase your branch on main.

File: src/Websites/Websites/help/Publish-AzWebApp.md [hunk 3]
```diff
@@ -197,6 +199,21 @@ Accept pipeline input: True (ByPropertyName)
 Accept wildcard characters: False
 ```
 
+### -ProgressAction
```
Comment (patelchandni) [inline src/Websites/Websites/help/Publish-AzWebApp.md:202]: I think ProgressAction is removed on "main" branch. Was this added back intentionally or just need to rebase your branch on main.

**Review by patelchandni — CHANGES_REQUESTED** [review]

File: src/Websites/Websites/help/Publish-AzWebApp.md [hunk 4]
```diff
@@ -15,15 +15,17 @@ Deploys an Azure Web App from a ZIP, JAR, or WAR file using zipdeploy.
...
 ```
 
 ### FromResourceName
 ```
 Publish-AzWebApp -ArchivePath <String> [-Type <String>] [-Clean] [-Async] [-Restart] [-TargetPath <String>]
- [-IgnoreStack] [-Reset] [-Force] [-AsJob] [-Timeout <Double>] [-ResourceGroupName] <String> [-Name] <String>
- [[-Slot] <String>] [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [<CommonParameters>]
+ [-IgnoreStack] [-Reset] [-UseZipDeploy] [-Force] [-AsJob] [-Timeout <Double>] [-ResourceGroupName] <String>
+ [-Name] <String> [[-Slot] <String>] [-DefaultProfile <IAzureContextContainer>]
+ [-ProgressAction <ActionPreference>] [-WhatIf] [-Confirm] [<CommonParameters>]
```
[the remark this answers is not in the document]
Comment (dannysongg) [reply src/Websites/Websites/help/Publish-AzWebApp.md:28]: This was automatically added when I regenerated the help.md. I will remove

File: src/Websites/Websites/ChangeLog.md [hunk 5]
```diff
@@ -18,6 +18,7 @@
         - Additional information about change #1
 -->
 ## Upcoming Release
+* `Publish-AzWebApp` now uses ZipDeploy to publish Function Apps
```
Comment (YanaXu) [inline src/Websites/Websites/ChangeLog.md:21]:
```suggestion
* Supported the parameter `ZipDeploy` in cmdlet `Publish-AzWebApp` to publish Function Apps
```

File: src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs [hunk 6]
```diff
@@ -86,74 +89,89 @@ public override void ExecuteCmdlet()
...
             var scmUrl = new Uri(user.ScmUri);
-            var deployUri = new Uri(scmUrl, apiPath);
-
-            var uriBuilder = new UriBuilder(deployUri);
-            var paramValues = HttpUtility.ParseQueryString(uriBuilder.Query);
+            string deployUrl;
 
-            string fileExtention = Path.GetExtension(ArchivePath);
-            string[] validTypes = { "war", "jar", "ear", "zip", "static" };
+            var appKind = WebsitesClient.GetWebApp(ResourceGroupName, Name, null).Kind;
```
Comment (YanaXu) [inline src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:94]: Please add tests for this update.

**Comment by YanaXu** [comment]
Please suppress the "Invalid_Cmdlet" errors.

**Pull request closed without merging**